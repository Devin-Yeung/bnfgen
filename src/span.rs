use miette::SourceSpan;

#[derive(Debug, Clone, Copy, Eq, PartialEq)]
pub struct Span {
    start: usize,
    end: usize,
}

impl Span {
    pub fn new(start: usize, end: usize) -> Self {
        Self { start, end }
    }
}

impl From<Span> for SourceSpan {
    fn from(val: Span) -> Self {
        SourceSpan::from(val.start..val.end)
    }
}

impl From<logos::Span> for Span {
    fn from(val: logos::Span) -> Self {
        Self {
            start: val.start,
            end: val.end,
        }
    }
}
