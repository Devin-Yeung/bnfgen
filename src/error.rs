use crate::span::Span;

pub type Result<T> = std::result::Result<T, Error>;

#[derive(thiserror::Error, miette::Diagnostic, Debug, Eq, PartialEq, Clone)]
pub enum Error {
    #[error("Unrecognized token")]
    UnrecognizedToken {
        #[label("expect {expect}")]
        span: Span,
        expect: String,
    },
    #[error("Unexpected extra token")]
    ExtraToken {
        #[label("this extra token is unexpected")]
        span: Span,
    },
    #[error("Unrecognized EOF")]
    UnrecognizedEof {
        #[label("expect {expect}")]
        span: Span,
        expect: String,
    },
    #[error("Undefined non-terminal")]
    UndefinedNonTerminal {
        #[label("this non-terminal is undefined")]
        span: Span,
    },
    #[error("Duplicated rules found")]
    DuplicatedRules {
        #[label("this rule is duplicated")]
        span: Span,
        #[label("previous defined here")]
        prev: Span,
    },
    #[error("Invalid repeat range")]
    InvalidRepeatRange {
        #[label("min should be less than or equal to max")]
        span: Span,
    },
    #[error("Found unreachable rules")]
    UnreachableRules {
        #[label(collection, "this rule is unreachable")]
        spans: Vec<Span>,
    },
    #[error("May be trapped in a dead loop")]
    TrapLoop {
        #[label(collection, "this rule may be trapped in a dead loop")]
        spans: Vec<Span>,
    },
    #[error(transparent)]
    #[diagnostic(transparent)]
    LexicalError(#[from] crate::token::LexicalError),
}
