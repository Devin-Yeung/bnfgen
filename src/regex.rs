use crate::error::{Error, Result};
use crate::span::Span;
use rand::Rng;
use regex_syntax::hir::{Class, Hir, HirKind};

#[repr(transparent)]
#[derive(Debug)]
pub struct Regex {
    hir: Hir,
}

impl Regex {
    fn new(input: &str) -> Self {
        let hir = regex_syntax::Parser::new().parse(input).unwrap();
        Self { hir }
    }

    pub fn spanned(input: &str, l: usize, r: usize) -> Result<Regex> {
        let hir = regex_syntax::Parser::new()
            .parse(input)
            .map_err(|_| Error::InvalidRegex {
                span: Span::new(l, r),
            })?;
        Ok(Regex { hir })
    }

    pub fn generate<R: Rng>(&self, rng: &mut R, terminals: &[&str]) -> String {
        // if regex produce a string that is a terminal, re-generate it
        loop {
            let s = Self::helper(&self.hir, rng);
            if !terminals.contains(&s.as_str()) {
                return s;
            }
        }
    }

    fn helper<R: Rng>(re: &Hir, rng: &mut R) -> String {
        match re.kind() {
            HirKind::Empty => String::new(),
            HirKind::Literal(lit) => String::from_utf8(lit.0.clone().into()).unwrap(),
            HirKind::Repetition(rep) => {
                let mut buf = Vec::new();
                // todo: allow manually set the max reps
                for _ in 0..rng.gen_range(rep.min..=rep.max.unwrap_or(5)) {
                    buf.push(Self::helper(&rep.sub, rng));
                }
                buf.join("")
            }
            HirKind::Concat(cat) => cat.iter().map(|h| Self::helper(h, rng)).collect(),
            HirKind::Alternation(alt) => {
                let idx = rng.gen_range(0..alt.len());
                Self::helper(&alt[idx], rng)
            }
            HirKind::Class(cls) => match cls {
                Class::Unicode(unicode) => {
                    let idx = rng.gen_range(0..unicode.iter().count());
                    let range = unicode.iter().nth(idx).unwrap();
                    let pick = rng.gen_range(range.start()..=range.end());
                    pick.to_string()
                }
                Class::Bytes(bytes) => {
                    let idx = rng.gen_range(0..bytes.iter().count());
                    let range = bytes.iter().nth(idx).unwrap();
                    let pick = rng.gen_range(range.start()..=range.end()) as char;
                    pick.to_string()
                }
            },
            HirKind::Look(_) => todo!(),
            HirKind::Capture(cap) => Self::helper(&cap.sub, rng),
        }
    }
}

#[cfg(test)]
mod test {
    use rand::rngs::StdRng;
    use rand::SeedableRng;

    #[test]
    fn it_works() {
        let mut rng = StdRng::seed_from_u64(42);
        let re = super::Regex::new("[a-zA-Z0-9]*");
        let generated = (0..10)
            .map(|_| re.generate(&mut rng, &["M"]))
            .collect::<Vec<_>>();
        insta::assert_debug_snapshot!(generated);
    }
}
