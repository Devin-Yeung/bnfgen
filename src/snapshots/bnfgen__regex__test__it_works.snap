---
source: src/regex.rs
expression: generated
---
[
    "",
    "5qk",
    "6Jeq4",
    "",
    "pWS8",
    "fA5",
    "6G",
    "MnB34",
    "",
    "FK6l",
]
